### PR TITLE
Attribute table optimization: do not load hidden fields until required

### PR DESCRIPTION
**Pull request by elpaso**
- defer hidden fields fetching
- allow caching of mixed full and subset attributes features

This greatly improves attribute table loading times when there are hidden virtual fields and (in a more limited way) when there are hidden non-virtual fields.

Funded by: QCooperative - QTIBIA Engineering

### CODE REVIEW

File: src/gui/attributetable/qgsdualview.cpp [hunk 1]
```diff
@@ -1142,6 +1143,36 @@ bool QgsDualView::modifySort()
...
+QgsAttributeList QgsDualView::requiredAttributes( const QgsVectorLayer *layer )
+{
+  QSet<int> attributes;
+
+  const QgsAttributeTableConfig config { layer->attributeTableConfig() };
+
+  const QVector<QgsAttributeTableConfig::ColumnConfig> constCols { config.columns() };
+  for ( int colIdx = 0; colIdx < constCols.count(); ++colIdx )
+  {
+    const QgsAttributeTableConfig::ColumnConfig col { constCols.at( colIdx ) };
```
Comment (troopa81) [inline src/gui/attributetable/qgsdualview.cpp:1155]:
```suggestion
 for ( const QgsAttributeTableConfig::ColumnConfig col :  constCols )
  {
```

**Review by troopa81 — APPROVED** [review]
few cosmetic comments, but the overall looks good!

File: src/gui/attributetable/qgsdualview.cpp [hunk 2]
```diff
@@ -1142,6 +1143,36 @@ bool QgsDualView::modifySort()
...
+    if ( col.type == QgsAttributeTableConfig::Type::Field && ! col.hidden )
+    {
+      attributes.insert( layer->fields().lookupField( col.name ) );
+    }
+  }
+
+  const QSet<int> colAttrs { attributes };
+  for ( const int attrIdx : std::as_const( colAttrs ) )
```
[the remark this answers is not in the document]
Comment (elpaso) [reply src/gui/attributetable/qgsdualview.cpp:1163]: Technically not, but it's a noop and makes it clear and explicit that the iteration won't detach.

I think it was discussed in the past and we decided to always use std::as_const even when it was redundant, I made it a rule after that discussion but I really can't recall when it was.

[the remark this answers is not in the document]
Comment (troopa81) [reply src/gui/attributetable/qgsdualview.cpp:1163]: OK

**Pull request merged**